op-node/rollup/engine/api.go

```diff
@@ -0,0 +1,124 @@
+package engine
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/rpc"
+
+	"github.com/ethereum-optimism/optimism/op-node/rollup/derive"
+	"github.com/ethereum-optimism/optimism/op-service/apis"
+	"github.com/ethereum-optimism/optimism/op-service/eth"
+	opsigner "github.com/ethereum-optimism/optimism/op-service/signer"
+)
+
+// RollupAPI is the API we serve as rollup-node to interact with the execution engine and forkchoice state.
+type RollupAPI interface {
+	apis.BuildAPI
+	apis.CommitAPI
+}
+
+var _ RollupAPI = (*EngineController)(nil)
+
+func (ec *EngineController) OpenBlock(ctx context.Context, parent eth.BlockID, attrs *eth.PayloadAttributes) (eth.PayloadInfo, error) {
+	ec.mu.Lock()
+	defer ec.mu.Unlock()
+
+	_, err := ec.engine.L2BlockRefByHash(ctx, parent.Hash)
+	if err != nil {
+		return eth.PayloadInfo{}, fmt.Errorf("failed to retrieve parent block %s from engine: %w", parent, err)
+	}
+
+	if err := ec.initializeUnknowns(ctx); err != nil {
+		return eth.PayloadInfo{}, fmt.Errorf("failed to initialize forkchoice pre-state: %w", err)
+	}
+
+	fc := eth.ForkchoiceState{
+		HeadBlockHash:      parent.Hash,
+		SafeBlockHash:      ec.safeHead.Hash,
+		FinalizedBlockHash: ec.finalizedHead.Hash,
+	}
+	id, errTyp, err := startPayload(ctx, ec.engine, fc, attrs)
+	if err != nil {
+		switch errTyp {
+		case BlockInsertTemporaryErr:
+			// RPC errors are not persistent block processing errors
+			return eth.PayloadInfo{}, fmt.Errorf("temporarily cannot insert new safe block: %w", err)
+		case BlockInsertPrestateErr:
+			return eth.PayloadInfo{}, fmt.Errorf("need reset to resolve pre-state problem: %w", err)
+		case BlockInsertPayloadErr:
+			return eth.PayloadInfo{}, fmt.Errorf("invalid payload attributes")
+		default:
+			return eth.PayloadInfo{}, fmt.Errorf("unknown error type %d: %w", errTyp, err)
+		}
+	}
+	return eth.PayloadInfo{
+		ID:        id,
+		Timestamp: uint64(attrs.Timestamp),
+	}, nil
+}
+
+func (ec *EngineController) CancelBlock(ctx context.Context, id eth.PayloadInfo) error {
```

Comment:
It looks to me like `CancelBlock` is currently the same as `SealBlock` but throwing away the `ExecutionPayloadEnvelope`? If it is intended to be the same, maybe we can reuse the same code, e.g. let CancelBlock just call SealBlock?

Reply:
There is the error-handling difference from the todo comment. Fixing that todo now.